### PR TITLE
XID Rumble

### PR DESCRIPTION
**Pull request by JayFoxRox**
Add rumble support in XID [hacky] so we don't assert all the time.

### CODE REVIEW

File: hw/xbox/xid.c [hunk 1]
```diff
@@ -235,12 +252,24 @@ static void usb_xid_handle_control(USBDevice *dev, USBPacket *p,
...
-        DPRINTF("xid SET_REPORT %x\n", value);
-        assert(false);
+        DPRINTF("xid SET_REPORT 0x%x\n", value);
+        if (value == 0x200) { /* output */
+            /* Read length, then the entire packet */
+            memcpy(&s->out_state, data, sizeof(s->out_state));
+            assert(s->out_state.length == sizeof(s->out_state));
+            assert(s->out_state.length <= length);
+            //FIXME: Check actuator endianess
+            printf("Set rumble power to 0x%x, 0x%x\n",
```
Comment (espes) [inline hw/xbox/xid.c]: this should be a DPRINTF

File: hw/xbox/xid.c [hunk 2]
```diff
@@ -215,19 +222,29 @@ static void usb_xid_handle_control(USBDevice *dev, USBPacket *p,
...
     }
 
     switch (request) {
     /* HID requests */
     case ClassInterfaceRequest | HID_GET_REPORT:
-        DPRINTF("xid GET_REPORT %x\n", value);
+        DPRINTF("xid GET_REPORT 0x%x\n", value);
         if (value == 0x100) { /* input */
             assert(s->in_state.bLength <= length);
+//            s->in_state.bReportId++; FIXME: Like this?
```
Comment (espes) [inline hw/xbox/xid.c]: ?

File: hw/xbox/xid.c [hunk 3]
```diff
@@ -215,19 +222,29 @@ static void usb_xid_handle_control(USBDevice *dev, USBPacket *p,
...
+    static unsigned int pid = 0;
+    static size_t bytes = 0;
+    DPRINTF("xid handle_control 0x%x 0x%x, packet %zu, byte %d\n", request, value, ++pid, bytes += length);
+#if 0
+    if ((request == USB_REQ_SET_CONFIGURATION) && (value == 1)) {
+        /* The OpenXDK code will try to set configuration 1 on every read */
+        DPRINTF("xid handled by configuration hack\n");
+        return;
+    }
+#endif
```
Comment (espes) [inline hw/xbox/xid.c]: Is this still true? Dead code is annoying

File: hw/xbox/xid.c (hunk 1 above)
Comment (JayFoxRox) [inline hw/xbox/xid.c]: I decided for printf so we actually have some form of output. If we had a proper input layer we could set the rumble power, however, for devices such as mice and keyboards there should be another indicator. This is as simple as it gets.

File: hw/xbox/xid.c (hunk 2 above)
Comment (JayFoxRox) [inline hw/xbox/xid.c]: I'm not sure if bReportId is just a counter. Hence the question.

File: hw/xbox/xid.c (hunk 3 above)
Comment (JayFoxRox) [inline hw/xbox/xid.c]: It's true for many compiled OpenXDK applications.
This hack should still work.

Keep in mind that I wrote this XID code back in mid 2014 or something and can't remember the specifics of this "hack".
At least it's a rather clean "hack" and we can still remove it at a later point

File: hw/xbox/xid.c (hunk 2 above)
Comment (espes) [inline hw/xbox/xid.c]: Stick something like that into the comment

File: hw/xbox/xid.c (hunk 3 above)
Comment (espes) [inline hw/xbox/xid.c]: Could you remove the hack from this PR then?

File: hw/xbox/xid.c [hunk 4]
```diff
@@ -215,19 +222,29 @@ static void usb_xid_handle_control(USBDevice *dev, USBPacket *p,
...
     int ret = usb_desc_handle_control(dev, p, request, value, index, length, data);
     if (ret >= 0) {
+        DPRINTF("xid handled by usb_desc_handle_control: %d\n",ret);
```
Comment (espes) [inline hw/xbox/xid.c]: nit: space after comma

**Comment by JayFoxRox** [comment]
Should be fine now?

**Comment by espes** [comment]
Thanks!

**Pull request merged**